lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -9639,6 +9639,86 @@ namespace Js
+        {
+            FunctionBody* parentFuncBody = pfuncCaller->GetFunctionBody();
+            ModuleID id = parentFuncBody->GetModuleID();
+
+            if (!JavascriptString::Is(specifier)) {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedString);
+            }
+
+            LPCOLESTR moduleName = Js::JavascriptString::FromVar(specifier)->GetSz();
+
+            HRESULT hr = 0;
+            if (id != kmodGlobal)
+            {
+                SourceTextModuleRecord *referenceModuleRecord = parentFuncBody->GetScriptContext()->GetLibrary()->GetModuleRecord(id);
+
+                ASYNC_HOST_OPERATION_START(scriptContext->GetThreadContext());
+                hr = scriptContext->GetHostScriptContext()->FetchImportedModule(referenceModuleRecord, moduleName, &moduleRecordBase);
+                ASYNC_HOST_OPERATION_END(scriptContext->GetThreadContext());
+            }
+            else
+            {
+                DWORD_PTR dwReferencingSourceContext = parentFuncBody->GetFunctionInfo()->GetFunctionProxy()->GetSourceContextInfo()->dwHostSourceContext;
+
+                ASYNC_HOST_OPERATION_START(scriptContext->GetThreadContext());
+                hr = scriptContext->GetHostScriptContext()->FetchImportedModuleFromScript(dwReferencingSourceContext, moduleName, &moduleRecordBase);
+                ASYNC_HOST_OPERATION_END(scriptContext->GetThreadContext());
+            }
+
+            if (FAILED(hr))
+            {
+                Js::JavascriptError * error = scriptContext->GetLibrary()->CreateURIError();
+                Var value = JavascriptString::NewCopySz(moduleName, scriptContext);
+                JavascriptOperators::OP_SetProperty(error, PropertyIds::message, value, scriptContext);
+                return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, error, scriptContext);
+            }
+
+            moduleRecord = SourceTextModuleRecord::FromHost(moduleRecordBase);
+
+            if (moduleRecord->WasEvaluated())
+            {
+                return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(true, moduleRecord->GetNamespace(), scriptContext, moduleRecord);
+            }
+            else if (moduleRecord->GetErrorObject() != nullptr)
+            {
+                return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, moduleRecord->GetErrorObject(), scriptContext, moduleRecord);
+            }
+            else
+            {
+                JavascriptPromiseCapability *promiseCapability = moduleRecord->GetPromiseCapability();
+                if (promiseCapability == nullptr)
+                {
+                    BEGIN_TRANSLATE_TO_HRESULT(static_cast<ExceptionType>(ExceptionType_OutOfMemory | ExceptionType_StackOverflow));
+                    promiseCapability = JavascriptPromise::NewPromiseCapability(scriptContext->GetLibrary()->GetPromiseConstructor(), scriptContext);
```

Comment:
Good idea. Updated in 2nd iteration.

---
In reply to: [114912538](https://github.com/Microsoft/ChakraCore/pull/2913#discussion_r114912538) [](ancestors = 114912538)